src/api/config.i

```diff
@@ -0,0 +1,33 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+
+
+// Pass in cmake configurations to swig
+#define USE_CUDA 1
+#define USE_CUDNN 1
+#define USE_OPENCL 0
+#define USE_PYTHON 1
+#define USE_MKLDNN 1
+#define USE_JAVA 0
+#define CUDNN_VERSION 7401
+
+// SINGA version
+#define SINGA_MAJOR_VERSION 1
```

Comment:
since the latest version has some bugs(basic autograd examples can't pass), we implement upon older version to confirm our code works. can integrated with new version after bugs fixed.

Reply:
pls report the bugs.

Reply:
In our server (at ncrg), I created a new anaconda python 3.6 enivorment and install singa 2.0 using "conda install -c nusdbsystem -c conda-forge singa=2.0.0=cudnn7.3.1_cuda10.0_py36"

It passed the test: python -c "from singa import tensor"
Also, it passed the old optimizer example: incubator-singa/example/cifar10/train.py can run and train successfully.

However, the incubator-singa/examples/autograd/resnet.py cannot run, while the output is:
Start intialization............
  0%|                                                                                  | 0/200 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "examples/autograd/resnet.py", line 247, in <module>
    x = model(tx)
  File "examples/autograd/resnet.py", line 155, in __call__
    x = self.conv1(x)
  File "/home/dcsysh/anaconda3/envs/singa2/lib/python3.6/site-packages/singa/autograd.py", line 939, in __call__
    self.device_check(x, self.W, self.b)
  File "/home/dcsysh/anaconda3/envs/singa2/lib/python3.6/site-packages/singa/autograd.py", line 656, in device_check
    if var.device.id() != x_dev_id:
AttributeError: 'NoneType' object has no attribute 'device'

Reply:
Updated on 1 August 2019:

Concerning the above error, I found that there is a different between the implementation of `class _BatchNorm2d(Operation):` in master branch and dist_new branch.

In autograd.py, both the master branch and dist_new branch has modified (or debugged) the conv2d and batchnorm operator, but they modified it differently. Meanwhile, both conv2d in the master branch and dist_new branch can train and reduce loss of mnist simple CNN, so there is no big problem. However, the batch normalization is a much more complex case, because it includes non-training variables that are running means and running variances.

In the master branch, the running means and running variances (non-training variables) are in the forward function: `def forward(self, x, scale, bias, running_mean, running_var):`
https://github.com/apache/incubator-singa/blob/master/python/singa/autograd.py#L1099

When I run the code using the master branch dockerfile, the error is as follows:
```
root@26c9db193eb0:~/incubator-singa/examples/autograd# python3 resnet.py
Start intialization............
  0%|                                                                                                      | 0/200 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "resnet.py", line 249, in <module>
    for p, g in autograd.backward(loss):
  File "/root/incubator-singa/build/python/singa/autograd.py", line 135, in backward
    % (len(op.src), len(dxs))
AssertionError: the number of src ops (=5) and dx (=3) not match
```
I think the error is because the running_mean and running_var are in the forward function input arguments but are not training variables, so there are supposed to be three src ops but finally found 5.

Meanwhile, the dist_new branch has modified the batchnorm function (commit 2b3a857 by user ubuntu on Apr14) by moving the input arguments running_mean and running_var into the initialization function:
`def __init__(self, handle, running_mean, running_var, name=None):`
`def forward(self, x, scale, bias):`
https://github.com/xuewanqi/incubator-singa/blob/dist_new/python/singa/autograd.py#L1096
This one can run successfully but I am not sure if it can train and reduce loss.

Next, I will try training the resnet with real dataset to see if it can reduce the loss.

Reply:
I have trained the dist_new branch resnet (because resnet has batch norm) with cifar10 dataset, and obtained 92.5% test accuracy in 100 iteration with data augmentation.

```
ubuntu@ip-172-31-27-25:~/incubator-singa/examples/autograd$ python3 resnet_realdata.py
Loading data file cifar-10-batches-py/data_batch_1
Loading data file cifar-10-batches-py/data_batch_2
Loading data file cifar-10-batches-py/data_batch_3
Loading data file cifar-10-batches-py/data_batch_4
Loading data file cifar-10-batches-py/data_batch_5
Loading data file cifar-10-batches-py/test_batch
Start intialization............
Epoch=0: 100%|███████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 2927.551146, training accuracy = 0.338068
Test accuracy = 0.441306
Epoch=1: 100%|███████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.59it/s]
Training loss = 2110.360374, training accuracy = 0.511984
Test accuracy = 0.606571
Epoch=2: 100%|███████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 1658.897868, training accuracy = 0.623199
Test accuracy = 0.645232
Epoch=3: 100%|███████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.64it/s]
Training loss = 1354.082412, training accuracy = 0.694442
Test accuracy = 0.731170
Epoch=4: 100%|███████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 1155.785529, training accuracy = 0.743478
Test accuracy = 0.761318
Epoch=5: 100%|███████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.59it/s]
Training loss = 1022.750388, training accuracy = 0.773668
Test accuracy = 0.741286
Epoch=6: 100%|███████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 945.400214, training accuracy = 0.790373
Test accuracy = 0.795072
Epoch=7: 100%|███████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 840.933215, training accuracy = 0.814441
Test accuracy = 0.810096
Epoch=8: 100%|███████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 765.215148, training accuracy = 0.830566
Test accuracy = 0.807091
Epoch=9: 100%|███████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 701.153867, training accuracy = 0.845951
Test accuracy = 0.822316
Epoch=10: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 666.267428, training accuracy = 0.853073
Test accuracy = 0.851162
Epoch=11: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 606.699607, training accuracy = 0.866817
Test accuracy = 0.770232
Epoch=12: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 564.226388, training accuracy = 0.875760
Test accuracy = 0.811599
Epoch=13: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.58it/s]
Training loss = 545.325170, training accuracy = 0.877621
Test accuracy = 0.856771
Epoch=14: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.59it/s]
Training loss = 497.799559, training accuracy = 0.889885
Test accuracy = 0.873798
Epoch=15: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.59it/s]
Training loss = 472.436701, training accuracy = 0.895927
Test accuracy = 0.864083
Epoch=16: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 442.924634, training accuracy = 0.901288
Test accuracy = 0.856070
Epoch=17: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 412.631328, training accuracy = 0.907810
Test accuracy = 0.868289
Epoch=18: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.59it/s]
Training loss = 391.663337, training accuracy = 0.911932
Test accuracy = 0.878906
Epoch=19: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 364.874849, training accuracy = 0.918774
Test accuracy = 0.874399
Epoch=20: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.59it/s]
Training loss = 352.975441, training accuracy = 0.920355
Test accuracy = 0.885517
Epoch=21: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.60it/s]
Training loss = 324.665358, training accuracy = 0.927157
Test accuracy = 0.878506
Epoch=22: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 314.736589, training accuracy = 0.929617
Test accuracy = 0.875701
Epoch=23: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 296.109193, training accuracy = 0.934399
Test accuracy = 0.878405
Epoch=24: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 276.580544, training accuracy = 0.937960
Test accuracy = 0.878906
Epoch=25: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.58it/s]
Training loss = 267.193980, training accuracy = 0.939481
Test accuracy = 0.867488
Epoch=26: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 239.712160, training accuracy = 0.946063
Test accuracy = 0.886118
Epoch=27: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 238.206897, training accuracy = 0.946803
Test accuracy = 0.899139
Epoch=28: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.58it/s]
Training loss = 229.405735, training accuracy = 0.947343
Test accuracy = 0.873698
Epoch=29: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 211.089237, training accuracy = 0.951865
Test accuracy = 0.896735
Epoch=30: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 198.430727, training accuracy = 0.954665
Test accuracy = 0.900040
Epoch=31: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.60it/s]
Training loss = 191.366123, training accuracy = 0.956086
Test accuracy = 0.899639
Epoch=32: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 182.107810, training accuracy = 0.959287
Test accuracy = 0.902544
Epoch=33: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 172.826398, training accuracy = 0.960767
Test accuracy = 0.893530
Epoch=34: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 169.613187, training accuracy = 0.961008
Test accuracy = 0.895933
Epoch=35: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 155.248689, training accuracy = 0.965709
Test accuracy = 0.896334
Epoch=36: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 153.824897, training accuracy = 0.965969
Test accuracy = 0.905549
Epoch=37: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.58it/s]
Training loss = 146.810095, training accuracy = 0.966649
Test accuracy = 0.887921
Epoch=38: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.60it/s]
Training loss = 135.051945, training accuracy = 0.969390
Test accuracy = 0.899639
Epoch=39: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 133.820405, training accuracy = 0.969950
Test accuracy = 0.897436
Epoch=40: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 135.417706, training accuracy = 0.969230
Test accuracy = 0.914964
Epoch=41: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 120.854997, training accuracy = 0.972371
Test accuracy = 0.905449
Epoch=42: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 116.340978, training accuracy = 0.973211
Test accuracy = 0.905048
Epoch=43: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 110.309411, training accuracy = 0.975292
Test accuracy = 0.909455
Epoch=44: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 111.145267, training accuracy = 0.974912
Test accuracy = 0.914163
Epoch=45: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.60it/s]
Training loss = 101.818090, training accuracy = 0.977693
Test accuracy = 0.903546
Epoch=46: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 94.800353, training accuracy = 0.978513
Test accuracy = 0.910958
Epoch=47: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 98.355502, training accuracy = 0.977513
Test accuracy = 0.904347
Epoch=48: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 96.644517, training accuracy = 0.978333
Test accuracy = 0.907252
Epoch=49: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 93.228892, training accuracy = 0.979253
Test accuracy = 0.908353
Epoch=50: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.60it/s]
Training loss = 88.646000, training accuracy = 0.980034
Test accuracy = 0.909355
Epoch=51: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 85.061176, training accuracy = 0.981314
Test accuracy = 0.906450
Epoch=52: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 84.551198, training accuracy = 0.980994
Test accuracy = 0.913962
Epoch=53: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 80.467703, training accuracy = 0.982634
Test accuracy = 0.909555
Epoch=54: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.64it/s]
Training loss = 75.081518, training accuracy = 0.983395
Test accuracy = 0.905449
Epoch=55: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.58it/s]
Training loss = 74.390348, training accuracy = 0.983635
Test accuracy = 0.921975
Epoch=56: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 71.364901, training accuracy = 0.984235
Test accuracy = 0.908754
Epoch=57: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 63.377296, training accuracy = 0.986216
Test accuracy = 0.896835
Epoch=58: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 64.011915, training accuracy = 0.985715
Test accuracy = 0.906250
Epoch=59: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 60.711218, training accuracy = 0.986156
Test accuracy = 0.910757
Epoch=60: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 60.320302, training accuracy = 0.986516
Test accuracy = 0.906751
Epoch=61: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.58it/s]
Training loss = 54.098782, training accuracy = 0.988016
Test accuracy = 0.890625
Epoch=62: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 64.635589, training accuracy = 0.986116
Test accuracy = 0.915966
Epoch=63: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 51.199125, training accuracy = 0.988516
Test accuracy = 0.905849
Epoch=64: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 55.281087, training accuracy = 0.988156
Test accuracy = 0.887620
Epoch=65: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 50.973392, training accuracy = 0.988596
Test accuracy = 0.919571
Epoch=66: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.59it/s]
Training loss = 52.769947, training accuracy = 0.988296
Test accuracy = 0.903746
Epoch=67: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.59it/s]
Training loss = 49.218939, training accuracy = 0.989437
Test accuracy = 0.885617
Epoch=68: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 54.984499, training accuracy = 0.987776
Test accuracy = 0.920573
Epoch=69: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 52.156214, training accuracy = 0.989277
Test accuracy = 0.911458
Epoch=70: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.60it/s]
Training loss = 44.693285, training accuracy = 0.990797
Test accuracy = 0.916266
Epoch=71: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 42.939453, training accuracy = 0.990557
Test accuracy = 0.916567
Epoch=72: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 44.035228, training accuracy = 0.990657
Test accuracy = 0.908554
Epoch=73: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.57it/s]
Training loss = 46.019535, training accuracy = 0.989917
Test accuracy = 0.915064
Epoch=74: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.64it/s]
Training loss = 38.398655, training accuracy = 0.991597
Test accuracy = 0.923478
Epoch=75: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.58it/s]
Training loss = 44.667590, training accuracy = 0.990497
Test accuracy = 0.906450
Epoch=76: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.60it/s]
Training loss = 40.122385, training accuracy = 0.991357
Test accuracy = 0.913562
Epoch=77: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.59it/s]
Training loss = 40.387087, training accuracy = 0.991157
Test accuracy = 0.919571
Epoch=78: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 40.673833, training accuracy = 0.991177
Test accuracy = 0.916466
Epoch=79: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 37.217760, training accuracy = 0.991357
Test accuracy = 0.892929
Epoch=80: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 38.898771, training accuracy = 0.991797
Test accuracy = 0.921575
Epoch=81: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 36.469006, training accuracy = 0.992298
Test accuracy = 0.908153
Epoch=82: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.60it/s]
Training loss = 42.442582, training accuracy = 0.990977
Test accuracy = 0.919972
Epoch=83: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.59it/s]
Training loss = 38.075628, training accuracy = 0.991677
Test accuracy = 0.915966
Epoch=84: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 38.809648, training accuracy = 0.991417
Test accuracy = 0.920974
Epoch=85: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 35.128319, training accuracy = 0.992298
Test accuracy = 0.919772
Epoch=86: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 29.161311, training accuracy = 0.993638
Test accuracy = 0.918169
Epoch=87: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.58it/s]
Training loss = 33.438135, training accuracy = 0.993138
Test accuracy = 0.917067
Epoch=88: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 30.297209, training accuracy = 0.993318
Test accuracy = 0.909856
Epoch=89: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.63it/s]
Training loss = 29.246781, training accuracy = 0.993998
Test accuracy = 0.914463
Epoch=90: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.60it/s]
Training loss = 33.667621, training accuracy = 0.992718
Test accuracy = 0.913562
Epoch=91: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.59it/s]
Training loss = 27.974487, training accuracy = 0.994018
Test accuracy = 0.917568
Epoch=92: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 28.299258, training accuracy = 0.993618
Test accuracy = 0.926583
Epoch=93: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 34.511342, training accuracy = 0.992177
Test accuracy = 0.919972
Epoch=94: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 30.856448, training accuracy = 0.993478
Test accuracy = 0.924179
Epoch=95: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 30.624649, training accuracy = 0.993198
Test accuracy = 0.927183
Epoch=96: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.60it/s]
Training loss = 24.563530, training accuracy = 0.994378
Test accuracy = 0.922075
Epoch=97: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 24.466792, training accuracy = 0.994958
Test accuracy = 0.920573
Epoch=98: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.61it/s]
Training loss = 28.265006, training accuracy = 0.994038
Test accuracy = 0.922276
Epoch=99: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [03:56<00:00,  6.62it/s]
Training loss = 27.351641, training accuracy = 0.993638
Test accuracy = 0.925481
ubuntu@ip-172-31-27-25:~/incubator-singa/examples/autograd$
```
The code used:
```
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

# the code is modified from
# https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py

try:
    import pickle
except ImportError:
    import cPickle as pickle

from singa import autograd
from singa import tensor
from singa import device
from singa import opt
import cv2
import numpy as np
from tqdm import trange

def load_dataset(filepath):
    print('Loading data file %s' % filepath)
    with open(filepath, 'rb') as fd:
        try:
            cifar10 = pickle.load(fd, encoding='latin1')
        except TypeError:
            cifar10 = pickle.load(fd)
    image = cifar10['data'].astype(dtype=np.uint8)
    image = image.reshape((-1, 3, 32, 32))
    label = np.asarray(cifar10['labels'], dtype=np.uint8)
    label = label.reshape(label.size, 1)
    return image, label


def load_train_data(dir_path='cifar-10-batches-py', num_batches=5):
    labels = []
    batchsize = 10000
    images = np.empty((num_batches * batchsize, 3, 32, 32), dtype=np.uint8)
    for did in range(1, num_batches + 1):
        fname_train_data = dir_path + "/data_batch_{}".format(did)
        image, label = load_dataset(fname_train_data)
        images[(did - 1) * batchsize:did * batchsize] = image
        labels.extend(label)
    images = np.array(images, dtype=np.float32)
    labels = np.array(labels, dtype=np.int32)
    return images, labels


def load_test_data(dir_path='cifar-10-batches-py'):
    images, labels = load_dataset(dir_path + "/test_batch")
    return np.array(images,  dtype=np.float32), np.array(labels, dtype=np.int32)

def normalize_for_resnet(train_x, test_x):   
    mean=[0.4914, 0.4822, 0.4465]
    std=[0.2023, 0.1994, 0.2010] 
    train_x /= 255
    test_x /= 255
    for ch in range(0,2):
        train_x[:, ch, :, :] -= mean[ch]
        train_x[:, ch, :, :] /= std[ch]
        test_x[:, ch, :, :] -= mean[ch]
        test_x[:, ch, :, :] /= std[ch]
    return train_x, test_x

def resize_dataset(x,IMG_SIZE):
    num_data = x.shape[0]
    dim = x.shape[1]
    X = np.zeros(shape=(num_data,dim,IMG_SIZE,IMG_SIZE), dtype=np.float32)
    for n in range(0,num_data):
        for d in range(0,dim):
            X[n, d, :, :] = cv2.resize(x[n , d, : ,:], (IMG_SIZE,IMG_SIZE)).astype(np.float32)
    return X

def augmentation(x, batch_size):
    xpad = np.pad(x, [[0, 0], [0, 0], [4, 4], [4, 4]], 'symmetric')
    for data_num in range(0, batch_size):
        offset = np.random.randint(8, size=2)
        x[data_num,:,:,:] = xpad[data_num, :, offset[0]: offset[0] + 32, offset[1]: offset[1] + 32]
        if_flip = np.random.randint(2)
        if (if_flip):
            x[data_num, :, :, :] = x[data_num, :, :, ::-1]
    return x

def accuracy(pred, target):
    y = np.argmax(pred, axis=1)
    t = np.argmax(target, axis=1)
    a = y == t
    return np.array(a, "int").sum()

def to_categorical(y, num_classes):
    y = np.array(y, dtype="int")
    n = y.shape[0]
    categorical = np.zeros((n, num_classes))
    for i in range(0,n):
      categorical[i, y[i]] = 1
      categorical = categorical.astype(np.float32)
    return categorical

if __name__ == '__main__':

    #load dataset
    #need to download with "/python3 incubator-singa/examples/cifar10/download_data.py py"
    train_x, train_y = load_train_data()
    test_x, test_y = load_test_data()
    train_x, test_x = normalize_for_resnet(train_x, test_x)

    from resnet import resnet50
    model = resnet50(num_classes=10)

    print('Start intialization............')
    dev = device.create_cuda_gpu_on(0)

    max_epoch = 100
    batch_size = 32
    IMG_SIZE = 224
    sgd = opt.SGD(lr=0.005, momentum=0.9, weight_decay=1e-5)
    tx = tensor.Tensor((batch_size, 3, IMG_SIZE, IMG_SIZE), dev, tensor.float32)
    ty = tensor.Tensor((batch_size,), dev, tensor.int32)
    num_train_batch = train_x.shape[0] // batch_size
    num_test_batch = test_x.shape[0] // batch_size
    idx = np.arange(train_x.shape[0], dtype=np.int32)


    for epoch in range(max_epoch):
        np.random.shuffle(idx)

        #Training Phase
        autograd.training = True
        train_correct = 0
        test_correct = 0
        train_loss = 0
        with trange(num_train_batch) as t:
            t.set_description('Epoch={}'.format(epoch))
            for b in t:
                x = train_x[idx[b * batch_size: (b + 1) * batch_size]]
                x = augmentation(x, batch_size)
                x = resize_dataset(x,IMG_SIZE)
                y = train_y[idx[b * batch_size: (b + 1) * batch_size]]
                tx.copy_from_numpy(x)
                ty.copy_from_numpy(y)
                out = model(tx)
                loss = autograd.softmax_cross_entropy(out, ty)               
                train_correct += accuracy(tensor.to_numpy(out), to_categorical(y, 10))
                train_loss += tensor.to_numpy(loss)[0]
                for p, g in autograd.backward(loss):
                    sgd.update(p, g)
                sgd.step()
        print('Training loss = %f, training accuracy = %f' % (train_loss, train_correct / (num_train_batch*(batch_size))))

        #Evaulation Phase
        autograd.training = False
        for b in range(num_test_batch):
            x = test_x[b * batch_size: (b + 1) * batch_size]
            x = resize_dataset(x,IMG_SIZE)
            y = test_y[b * batch_size: (b + 1) * batch_size]
            tx.copy_from_numpy(x)
            ty.copy_from_numpy(y)
            out_test = model(tx)
            test_correct += accuracy(tensor.to_numpy(out_test), to_categorical(y, 10))
        print('Test accuracy = %f' % (test_correct / (num_test_batch*(batch_size))))
```

Reply:
Next, I will try to train the resnet with cifar 10 using 8 gpu.

Reply:
Updated on 6th August: I removed a bug in the commit 0616000 which concerns to the number of parameters in the all-reduce. Then I did a 8 * K80 multi-GPUs training and evaluation test with a mnist dataset on simple CNN. It reduces the training loss from 802.7 to 42.2 in about 30 Epochs: 
```
Epoch=0: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 92.86it/s]Training loss = 802.659485, training accuracy = 0.713825
Test accuracy = 0.920025
Epoch=1: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 93.42it/s]Training loss = 246.589371, training accuracy = 0.916767
Test accuracy = 0.956106
Epoch=2: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 94.04it/s]Training loss = 175.012894, training accuracy = 0.941106
Test accuracy = 0.967208
Epoch=3: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 95.66it/s] Training loss = 144.684052, training accuracy = 0.951539
Test accuracy = 0.970806
Epoch=4: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.59it/s]Training loss = 120.399704, training accuracy = 0.959402
Test accuracy = 0.976049
Epoch=5: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.79it/s]Training loss = 107.832191, training accuracy = 0.963709
Test accuracy = 0.975946
Epoch=6: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.70it/s]Training loss = 96.289490, training accuracy = 0.967014
Test accuracy = 0.979441
Epoch=7: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.34it/s]Training loss = 88.031815, training accuracy = 0.970436
Test accuracy = 0.980983
Epoch=8: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 101.81it/s]Training loss = 79.349884, training accuracy = 0.973090
Test accuracy = 0.980058
Epoch=9: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 101.82it/s]Training loss = 77.825607, training accuracy = 0.974342
Test accuracy = 0.977282
Epoch=10: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 101.97it/s]Training loss = 74.710297, training accuracy = 0.974576
Test accuracy = 0.983861
Epoch=11: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 101.98it/s]Training loss = 69.400230, training accuracy = 0.976162
Test accuracy = 0.982936
Epoch=12: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.03it/s]Training loss = 65.100449, training accuracy = 0.978148
Test accuracy = 0.983553
Epoch=13: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.17it/s]Training loss = 65.113991, training accuracy = 0.978249
Test accuracy = 0.986534
Epoch=14: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 101.83it/s]Training loss = 63.065636, training accuracy = 0.978566
Test accuracy = 0.984683
Epoch=15: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.11it/s]Training loss = 58.334709, training accuracy = 0.980018
Test accuracy = 0.983758
Epoch=16: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.16it/s]Training loss = 58.280094, training accuracy = 0.980285
Test accuracy = 0.983655
Epoch=17: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.15it/s]Training loss = 53.226196, training accuracy = 0.981420
Test accuracy = 0.985197
Epoch=18: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.15it/s]Training loss = 55.968140, training accuracy = 0.980786
Test accuracy = 0.982422
Epoch=19: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.14it/s]Training loss = 52.761921, training accuracy = 0.982489
Test accuracy = 0.985814
Epoch=20: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 101.86it/s]Training loss = 51.989666, training accuracy = 0.982973
Test accuracy = 0.983758
Epoch=21: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 101.91it/s]Training loss = 52.571381, training accuracy = 0.982455
Test accuracy = 0.987973
Epoch=22: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 101.99it/s]Training loss = 49.347313, training accuracy = 0.983140
Test accuracy = 0.986637
Epoch=23: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 101.93it/s]Training loss = 49.053402, training accuracy = 0.983674
Test accuracy = 0.985814
Epoch=24: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 99.28it/s] Training loss = 46.263908, training accuracy = 0.984442
Test accuracy = 0.986431
Epoch=25: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 104.22it/s]Training loss = 46.021286, training accuracy = 0.984275
Test accuracy = 0.987664
Epoch=26: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 103.67it/s]Training loss = 45.950298, training accuracy = 0.984091
Test accuracy = 0.986534
Epoch=27: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.87it/s]Training loss = 43.926952, training accuracy = 0.984675
Test accuracy = 0.987150
Epoch=28: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.89it/s]Training loss = 44.020412, training accuracy = 0.985110
Test accuracy = 0.983450
Epoch=29: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 103.06it/s]Training loss = 41.906254, training accuracy = 0.985744
Test accuracy = 0.984375
Epoch=30: 100%|¦¦¦¦¦¦¦¦¦¦| 117/117 [00:01<00:00, 102.93it/s]Training loss = 42.237778, training accuracy = 0.985527
Test accuracy = 0.987664
```

The following is the code used:
```
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

# the code is modified from
# https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py

try:
    import pickle
except ImportError:
    import cPickle as pickle

from singa import singa_wrap as singa
from singa import autograd
from singa import tensor
from singa import device
from singa import opt
import numpy as np
from tqdm import trange

import os
import urllib.request
import gzip
import codecs

def load_dataset():
    train_x_url = 'http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz'
    train_y_url = 'http://yann.lecun.com/exdb/mnist/train-labels-idx1-ubyte.gz'
    valid_x_url = 'http://yann.lecun.com/exdb/mnist/t10k-images-idx3-ubyte.gz'
    valid_y_url = 'http://yann.lecun.com/exdb/mnist/t10k-labels-idx1-ubyte.gz'
    train_x = read_image_file(check_exist_or_download(train_x_url)).astype(
        np.float32)
    train_y = read_label_file(check_exist_or_download(train_y_url)).astype(
        np.float32)
    valid_x = read_image_file(check_exist_or_download(valid_x_url)).astype(
        np.float32)
    valid_y = read_label_file(check_exist_or_download(valid_y_url)).astype(
        np.float32)
    return train_x, train_y, valid_x, valid_y


def check_exist_or_download(url):

    download_dir = '/tmp/'

    name = url.rsplit('/', 1)[-1]
    filename = os.path.join(download_dir, name)
    if not os.path.isfile(filename):
        print("Downloading %s" % url)
        urllib.request.urlretrieve(url, filename)
    return filename


def read_label_file(path):
    with gzip.open(path, 'rb') as f:
        data = f.read()
        assert get_int(data[:4]) == 2049
        length = get_int(data[4:8])
        parsed = np.frombuffer(data, dtype=np.uint8, offset=8).reshape(
            (length))
        return parsed


def get_int(b):
    return int(codecs.encode(b, 'hex'), 16)


def read_image_file(path):
    with gzip.open(path, 'rb') as f:
        data = f.read()
        assert get_int(data[:4]) == 2051
        length = get_int(data[4:8])
        num_rows = get_int(data[8:12])
        num_cols = get_int(data[12:16])
        parsed = np.frombuffer(data, dtype=np.uint8, offset=16).reshape(
            (length, 1, num_rows, num_cols))
        return parsed



def normalize_for_resnet(train_x, test_x):   
    mean=[0.4914, 0.4822, 0.4465]
    std=[0.2023, 0.1994, 0.2010] 
    train_x /= 255
    test_x /= 255
    for ch in range(0,2):
        train_x[:, ch, :, :] -= mean[ch]
        train_x[:, ch, :, :] /= std[ch]
        test_x[:, ch, :, :] -= mean[ch]
        test_x[:, ch, :, :] /= std[ch]
    return train_x, test_x


def augmentation(x, batch_size):
    xpad = np.pad(x, [[0, 0], [0, 0], [4, 4], [4, 4]], 'symmetric')
    for data_num in range(0, batch_size):
        offset = np.random.randint(8, size=2)
        x[data_num,:,:,:] = xpad[data_num, :, offset[0]: offset[0] + 28, offset[1]: offset[1] + 28]
        if_flip = np.random.randint(2)
        if (if_flip):
            x[data_num, :, :, :] = x[data_num, :, :, ::-1]
    return x

def accuracy(pred, target):
    y = np.argmax(pred, axis=1)
    t = np.argmax(target, axis=1)
    a = y == t
    return np.array(a, "int").sum()

def to_categorical(y, num_classes):
    """
    Converts a class vector (integers) to binary class matrix.

    Args
        y: class vector to be converted into a matrix
            (integers from 0 to num_classes).
        num_classes: total number of classes.

    Return
        A binary matrix representation of the input.
    """
    y = np.array(y, dtype="int")
    n = y.shape[0]
    categorical = np.zeros((n, num_classes))
    categorical[np.arange(n), y] = 1
    categorical = categorical.astype(np.float32)
    return categorical

def accuracy(pred, target):
    y = np.argmax(pred, axis=1)
    t = np.argmax(target, axis=1)
    a = y == t
    return np.array(a, "int").sum()

class CNN:
    def __init__(self):
        self.conv1 = autograd.Conv2d(1, 20, 5, padding=0)
        self.conv2 = autograd.Conv2d(20, 50, 5, padding=0)
        self.linear1 = autograd.Linear(4 * 4 * 50, 500)
        self.linear2 = autograd.Linear(500, 10)
        self.pooling1 = autograd.MaxPool2d(2, 2, padding=0)
        self.pooling2 = autograd.MaxPool2d(2, 2, padding=0)

    def forward(self, x):
        y = self.conv1(x)
        y = autograd.relu(y)
        y = self.pooling1(y)
        y = self.conv2(y)
        y = autograd.relu(y)
        y = self.pooling2(y)
        y = autograd.flatten(y)
        y = self.linear1(y)
        y = autograd.relu(y)
        y = self.linear2(y)
        return y

def data_partition(dataset_x, dataset_y, rank_in_global, world_size):
    data_per_rank = dataset_x.shape[0] // world_size
    idx_start = rank_in_global * data_per_rank
    idx_end = (rank_in_global + 1) * data_per_rank
    return dataset_x[idx_start: idx_end], dataset_y[idx_start: idx_end]

def sychronize(tensor, dist_opt):
    singa.synch(tensor.data, dist_opt.communicator)
    # cannot use tensor/=dist_opt.world_size because "/=" not in place, but "-=" is in place
    tensor -= (dist_opt.world_size - 1) * tensor / dist_opt.world_size    



if __name__ == '__main__':


    sgd = opt.SGD(lr=0.04, momentum=0.9, weight_decay=1e-5)
    sgd = opt.DistOpt(sgd)

    # load data
    train_x, train_y, test_x, test_y = load_dataset()
    # normalization
    train_x = train_x / 255
    test_x = test_x / 255
    num_classes=10

    train_y = to_categorical(train_y, num_classes)
    test_y = to_categorical(test_y, num_classes)


    train_x, train_y = data_partition(train_x, train_y, sgd.rank_in_global, sgd.world_size)
    test_x, test_y = data_partition(test_x, test_y, sgd.rank_in_global, sgd.world_size)

    #print(train_y[0])

    print(np.shape(train_x))
    print(np.shape(train_y))

    # create model
    model = CNN()

    print('Start intialization............')
    dev = device.create_cuda_gpu_on(sgd.rank_in_local)

    max_epoch = 100
    batch_size = 64
    IMG_SIZE = 28
    tx = tensor.Tensor((batch_size, 1, IMG_SIZE, IMG_SIZE), dev, tensor.float32)
    ty = tensor.Tensor((batch_size, num_classes), dev, tensor.int32)
    num_train_batch = train_x.shape[0] // batch_size
    num_test_batch = test_x.shape[0] // batch_size
    idx = np.arange(train_x.shape[0], dtype=np.int32)
    reducer = tensor.Tensor((1,), dev, tensor.float32)

    #allreduce the initialize parameter
    autograd.training = True
    #x = np.zeros(shape=[batch_size, 1, IMG_SIZE, IMG_SIZE], dtype=np.float32)
    #y = np.zeros(shape=[batch_size], dtype=np.int32)
    x = np.random.randn(batch_size, 1, IMG_SIZE, IMG_SIZE).astype(np.float32)
    y = np.zeros( shape=(batch_size, num_classes), dtype=np.int32)
    tx.copy_from_numpy(x)
    ty.copy_from_numpy(y)
    out = model.forward(tx)
    loss = autograd.softmax_cross_entropy(out, ty)               
    for p, g in autograd.backward(loss):
        #p=sgd.all_reduce(p)
        sychronize(p, sgd)

    for epoch in range(max_epoch):
        np.random.shuffle(idx)

        #Training Phase
        autograd.training = True
        train_correct = np.zeros(shape=[1],dtype=np.float32)
        test_correct = np.zeros(shape=[1],dtype=np.float32)
        train_loss = np.zeros(shape=[1],dtype=np.float32)
        with trange(num_train_batch) as t:
            t.set_description('Epoch={}'.format(epoch))
            for b in t:
                x = train_x[idx[b * batch_size: (b + 1) * batch_size]]
                x = augmentation(x, batch_size)
                y = train_y[idx[b * batch_size: (b + 1) * batch_size]]
                tx.copy_from_numpy(x)
                ty.copy_from_numpy(y)
                out = model.forward(tx)
                loss = autograd.softmax_cross_entropy(out, ty)               
                train_correct += accuracy(tensor.to_numpy(out), y)
                train_loss += tensor.to_numpy(loss)[0]
                for p, g in autograd.backward(loss):
                    sgd.update(p, g)

        #reduce the accuracy from multiple device
        reducer.copy_from_numpy(train_correct)
        reducer=sgd.all_reduce(reducer)
        train_correct = tensor.to_numpy(reducer) 

        #reduce the loss from multiple device
        reducer.copy_from_numpy(train_loss)
        reducer=sgd.all_reduce(reducer)
        train_loss = tensor.to_numpy(reducer) * sgd.world_size

        if(sgd.rank_in_global==0):
            print('Training loss = %f, training accuracy = %f' % (train_loss, train_correct / (num_train_batch*batch_size)), flush=True)

        #Evaulation Phase
        autograd.training = False
        for b in range(num_test_batch):
            x = test_x[b * batch_size: (b + 1) * batch_size]
            y = test_y[b * batch_size: (b + 1) * batch_size]
            tx.copy_from_numpy(x)
            ty.copy_from_numpy(y)
            out_test = model.forward(tx)
            test_correct += accuracy(tensor.to_numpy(out_test), y)

        reducer.copy_from_numpy(test_correct)
        reducer=sgd.all_reduce(reducer)
        test_correct = tensor.to_numpy(reducer) 

        if(sgd.rank_in_global==0):
            print('Test accuracy = %f' % (test_correct / (num_test_batch*(batch_size))), flush=True)
```

Reply:
In additional to the above, I also did a 8 * K80 multi-GPUs training and evaluation test with a CIFAR-10 dataset on resnet 50. It reduces the training loss from 3983.8 to 345.7 in about 30 Epochs, and evaluation accuracy to 86.8%. However, this does not include the synchronization of running mean and variance before the evaluation phase:
```
Epoch=0: 100%|██████████| 195/195 [06:06<00:00,  1.91s/it]Training loss = 3983.820557, training accuracy = 0.225260
Test accuracy = 0.347556
Epoch=1: 100%|██████████| 195/195 [06:17<00:00,  1.94s/it]Training loss = 2628.622070, training accuracy = 0.379768
Test accuracy = 0.437700
Epoch=2: 100%|██████████| 195/195 [06:12<00:00,  1.89s/it]Training loss = 2347.072266, training accuracy = 0.448558
Test accuracy = 0.459936
Epoch=3: 100%|██████████| 195/195 [06:13<00:00,  1.88s/it]Training loss = 2075.987305, training accuracy = 0.517348
Test accuracy = 0.548978
Epoch=4: 100%|██████████| 195/195 [06:19<00:00,  1.97s/it]Training loss = 1890.109985, training accuracy = 0.566847
Test accuracy = 0.594451
Epoch=5: 100%|██████████| 195/195 [06:13<00:00,  1.92s/it]Training loss = 1720.395142, training accuracy = 0.606911
Test accuracy = 0.633413
Epoch=6: 100%|██████████| 195/195 [06:10<00:00,  1.92s/it]Training loss = 1555.737549, training accuracy = 0.645753
Test accuracy = 0.659054
Epoch=7: 100%|██████████| 195/195 [06:14<00:00,  1.91s/it]Training loss = 1385.688477, training accuracy = 0.687220
Test accuracy = 0.709836
Epoch=8: 100%|██████████| 195/195 [06:20<00:00,  1.97s/it]Training loss = 1269.426270, training accuracy = 0.714523
Test accuracy = 0.735477
Epoch=9: 100%|██████████| 195/195 [06:15<00:00,  1.91s/it]Training loss = 1137.953979, training accuracy = 0.746054
Test accuracy = 0.745393
Epoch=10: 100%|██████████| 195/195 [06:11<00:00,  1.88s/it]Training loss = 1031.773071, training accuracy = 0.770353
Test accuracy = 0.750501
Epoch=11: 100%|██████████| 195/195 [06:10<00:00,  1.89s/it]Training loss = 956.600037, training accuracy = 0.788261
Test accuracy = 0.777744
Epoch=12: 100%|██████████| 195/195 [06:16<00:00,  1.92s/it]Training loss = 881.050171, training accuracy = 0.804167
Test accuracy = 0.793369
Epoch=13: 100%|██████████| 195/195 [06:16<00:00,  1.92s/it]Training loss = 828.298828, training accuracy = 0.818309
Test accuracy = 0.807692
Epoch=14: 100%|██████████| 195/195 [06:11<00:00,  1.90s/it]Training loss = 790.558838, training accuracy = 0.823918
Test accuracy = 0.795373
Epoch=15: 100%|██████████| 195/195 [06:13<00:00,  1.90s/it]Training loss = 740.679871, training accuracy = 0.833734
Test accuracy = 0.816707
Epoch=16: 100%|██████████| 195/195 [06:20<00:00,  1.95s/it]Training loss = 691.391479, training accuracy = 0.846855
Test accuracy = 0.818510
Epoch=17: 100%|██████████| 195/195 [06:16<00:00,  1.89s/it]Training loss = 657.708130, training accuracy = 0.853986
Test accuracy = 0.826122
Epoch=18: 100%|██████████| 195/195 [06:10<00:00,  1.88s/it]Training loss = 627.918579, training accuracy = 0.860216
Test accuracy = 0.844752
Epoch=19: 100%|██████████| 195/195 [06:13<00:00,  1.91s/it]Training loss = 592.768982, training accuracy = 0.869551
Test accuracy = 0.845653
Epoch=20: 100%|██████████| 195/195 [06:19<00:00,  1.97s/it]Training loss = 561.560608, training accuracy = 0.875060
Test accuracy = 0.835938
Epoch=21: 100%|██████████| 195/195 [06:15<00:00,  1.97s/it]Training loss = 533.083740, training accuracy = 0.881370
Test accuracy = 0.849860
Epoch=22: 100%|██████████| 195/195 [06:12<00:00,  1.91s/it]Training loss = 508.004578, training accuracy = 0.885056
Test accuracy = 0.833434
Epoch=23: 100%|██████████| 195/195 [06:12<00:00,  1.92s/it]Training loss = 477.516602, training accuracy = 0.892488
Test accuracy = 0.858474
Epoch=24: 100%|██████████| 195/195 [06:20<00:00,  1.96s/it]Training loss = 455.839996, training accuracy = 0.896595
Test accuracy = 0.867388
Epoch=25: 100%|██████████| 195/195 [06:16<00:00,  1.95s/it]Training loss = 434.568390, training accuracy = 0.904327
Test accuracy = 0.858774
Epoch=26: 100%|██████████| 195/195 [06:10<00:00,  1.87s/it]Training loss = 414.232391, training accuracy = 0.907071
Test accuracy = 0.833333
Epoch=27: 100%|██████████| 195/195 [06:13<00:00,  1.87s/it]Training loss = 400.625458, training accuracy = 0.909275
Test accuracy = 0.858974
Epoch=28: 100%|██████████| 195/195 [06:20<00:00,  1.95s/it]Training loss = 378.750885, training accuracy = 0.914443
Test accuracy = 0.865885
Epoch=29: 100%|██████████| 195/195 [06:14<00:00,  1.91s/it]Training loss = 369.449249, training accuracy = 0.917548
Test accuracy = 0.871394
Epoch=30: 100%|██████████| 195/195 [06:13<00:00,  1.93s/it]Training loss = 345.693939, training accuracy = 0.921935
Test accuracy = 0.868389
```
The code used is as below:
```
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

try:
    import pickle
except ImportError:
    import cPickle as pickle
    
from singa import singa_wrap as singa
from singa import autograd
from singa import tensor
from singa import device
from singa import opt
import cv2
#from scipy import misc
import numpy as np
from tqdm import trange

def load_dataset(filepath):
    print('Loading data file %s' % filepath)
    with open(filepath, 'rb') as fd:
        try:
            cifar10 = pickle.load(fd, encoding='latin1')
        except TypeError:
            cifar10 = pickle.load(fd)
    image = cifar10['data'].astype(dtype=np.uint8)
    image = image.reshape((-1, 3, 32, 32))
    label = np.asarray(cifar10['labels'], dtype=np.uint8)
    label = label.reshape(label.size, 1)
    return image, label


def load_train_data(dir_path='cifar-10-batches-py', num_batches=5):
    labels = []
    batchsize = 10000
    images = np.empty((num_batches * batchsize, 3, 32, 32), dtype=np.uint8)
    for did in range(1, num_batches + 1):
        fname_train_data = dir_path + "/data_batch_{}".format(did)
        image, label = load_dataset(fname_train_data)
        images[(did - 1) * batchsize:did * batchsize] = image
        labels.extend(label)
    images = np.array(images, dtype=np.float32)
    labels = np.array(labels, dtype=np.int32)
    return images, labels


def load_test_data(dir_path='cifar-10-batches-py'):
    images, labels = load_dataset(dir_path + "/test_batch")
    return np.array(images,  dtype=np.float32), np.array(labels, dtype=np.int32)

def normalize_for_resnet(train_x, test_x):   
    mean=[0.4914, 0.4822, 0.4465]
    std=[0.2023, 0.1994, 0.2010] 
    train_x /= 255
    test_x /= 255
    for ch in range(0,2):
        train_x[:, ch, :, :] -= mean[ch]
        train_x[:, ch, :, :] /= std[ch]
        test_x[:, ch, :, :] -= mean[ch]
        test_x[:, ch, :, :] /= std[ch]
    return train_x, test_x

def resize_dataset(x,IMG_SIZE):
    num_data = x.shape[0]
    dim = x.shape[1]
    X = np.zeros(shape=(num_data,dim,IMG_SIZE,IMG_SIZE), dtype=np.float32)
    for n in range(0,num_data):
        for d in range(0,dim):
            X[n, d, :, :] = cv2.resize(x[n , d, : ,:], (IMG_SIZE,IMG_SIZE)).astype(np.float32)
    return X

def augmentation(x, batch_size):
    xpad = np.pad(x, [[0, 0], [0, 0], [4, 4], [4, 4]], 'symmetric')
    for data_num in range(0, batch_size):
        offset = np.random.randint(8, size=2)
        x[data_num,:,:,:] = xpad[data_num, :, offset[0]: offset[0] + 32, offset[1]: offset[1] + 32]
        if_flip = np.random.randint(2)
        if (if_flip):
            x[data_num, :, :, :] = x[data_num, :, :, ::-1]
    return x

def accuracy(pred, target):
    y = np.argmax(pred, axis=1)
    t = np.argmax(target, axis=1)
    a = y == t
    return np.array(a, "int").sum()

def to_categorical(y, num_classes):
    y = np.array(y, dtype="int")
    n = y.shape[0]
    categorical = np.zeros((n, num_classes))
    for i in range(0,n):
      categorical[i, y[i]] = 1
      categorical = categorical.astype(np.float32)
    return categorical

def data_partition(dataset_x, dataset_y, rank_in_global, world_size):
    data_per_rank = dataset_x.shape[0] // world_size
    idx_start = rank_in_global * data_per_rank
    idx_end = (rank_in_global + 1) * data_per_rank
    return dataset_x[idx_start: idx_end], dataset_y[idx_start: idx_end]

def sychronize(tensor, dist_opt):
    singa.synch(tensor.data, dist_opt.communicator)
    # cannot use tensor/=dist_opt.world_size because "/=" not in place, but "-=" is in place
    tensor -= (dist_opt.world_size - 1) * tensor / dist_opt.world_size    

if __name__ == '__main__':


    sgd = opt.SGD(lr=0.04, momentum=0.9, weight_decay=1e-5)
    sgd = opt.DistOpt(sgd)

    #load dataset
    #need to download with "/python3 incubator-singa/examples/cifar10/download_data.py py"
    train_x, train_y = load_train_data()
    test_x, test_y = load_test_data()
    train_x, test_x = normalize_for_resnet(train_x, test_x)
    train_x, train_y = data_partition(train_x, train_y, sgd.rank_in_global, sgd.world_size)
    test_x, test_y = data_partition(test_x, test_y, sgd.rank_in_global, sgd.world_size)

    num_classes=10

    from resnet import resnet50
    model = resnet50(num_classes=num_classes)

    print('Start intialization............')
    dev = device.create_cuda_gpu_on(sgd.rank_in_local)

    max_epoch = 100
    batch_size = 32
    IMG_SIZE = 224
    tx = tensor.Tensor((batch_size, 3, IMG_SIZE, IMG_SIZE), dev, tensor.float32)
    ty = tensor.Tensor((batch_size,), dev, tensor.int32)
    num_train_batch = train_x.shape[0] // batch_size
    num_test_batch = test_x.shape[0] // batch_size
    idx = np.arange(train_x.shape[0], dtype=np.int32)
    reducer = tensor.Tensor((1,), dev, tensor.float32)

    #allreduce the initialize parameter
    autograd.training = True
    #x = np.zeros(shape=[batch_size, 3, IMG_SIZE, IMG_SIZE], dtype=np.float32)
    #y = np.zeros(shape=[batch_size], dtype=np.int32)
    x = np.random.randn(batch_size, 3, IMG_SIZE, IMG_SIZE).astype(np.float32)
    y = np.random.randint(0, num_classes, batch_size, dtype=np.int32)
    tx.copy_from_numpy(x)
    ty.copy_from_numpy(y)
    out = model(tx)
    loss = autograd.softmax_cross_entropy(out, ty)               
    for p, g in autograd.backward(loss):
        sychronize(p, sgd)

    for epoch in range(max_epoch):
        np.random.shuffle(idx)

        #Training Phase
        autograd.training = True
        train_correct = np.zeros(shape=[1],dtype=np.float32)
        test_correct = np.zeros(shape=[1],dtype=np.float32)
        train_loss = np.zeros(shape=[1],dtype=np.float32)
        with trange(num_train_batch) as t:
            t.set_description('Epoch={}'.format(epoch))
            for b in t:
                x = train_x[idx[b * batch_size: (b + 1) * batch_size]]
                x = augmentation(x, batch_size)
                x = resize_dataset(x,IMG_SIZE)
                y = train_y[idx[b * batch_size: (b + 1) * batch_size]]
                tx.copy_from_numpy(x)
                ty.copy_from_numpy(y)
                out = model(tx)
                loss = autograd.softmax_cross_entropy(out, ty)               
                train_correct += accuracy(tensor.to_numpy(out), to_categorical(y, num_classes)).astype(np.float32)
                train_loss += tensor.to_numpy(loss)[0]
                for p, g in autograd.backward(loss):
                    sgd.update(p, g)

        #print("rank"+str(sgd.rank_in_global)+": Acc="+str(train_correct)+". Loss="+str(train_loss), flush=True)

        #print("world size="+str(sgd.world_size), flush=True)

        #reduce all the accuracy and loss from different rank
        reducer.copy_from_numpy(train_correct)
        reducer=sgd.all_reduce(reducer)
        train_correct = tensor.to_numpy(reducer) 

        reducer.copy_from_numpy(train_loss)
        reducer=sgd.all_reduce(reducer)
        train_loss = tensor.to_numpy(reducer) * sgd.world_size

        #if(sgd.rank_in_global==0):
        #    print('Training loss = %f, Acc count = %f' % (train_loss, train_correct), flush=True)

        if(sgd.rank_in_global==0):
            print('Training loss = %f, training accuracy = %f' % (train_loss, train_correct / (num_train_batch*batch_size)), flush=True)


        #Evaulation Phase
        autograd.training = False
        for b in range(num_test_batch):
            x = test_x[b * batch_size: (b + 1) * batch_size]
            x = resize_dataset(x,IMG_SIZE)
            y = test_y[b * batch_size: (b + 1) * batch_size]
            tx.copy_from_numpy(x)
            ty.copy_from_numpy(y)
            out_test = model(tx)
            test_correct += accuracy(tensor.to_numpy(out_test), to_categorical(y, num_classes))

        reducer.copy_from_numpy(test_correct)
        reducer=sgd.all_reduce(reducer)
        test_correct = tensor.to_numpy(reducer) 

        if(sgd.rank_in_global==0):
            print('Test accuracy = %f' % (test_correct / (num_test_batch*(batch_size))), flush=True)

```

Reply:
From the above, we can now train simple CNN (MNIST dataset) and resnet (CIFAR-10 dataset). The remaining task is the synchronization of the running mean and variance. 
I tried to put the running mean and var in the batch return list of backward 

```
    def backward(self, dy):
        assert training is True and hasattr(
            self, "cache"
        ), "Please set training as True before do BP. "

        x, scale, mean, var = self.cache
        if isinstance(self.handle, singa.CudnnBatchNormHandle):
            dx, ds, db = singa.GpuBatchNormBackward(
                self.handle, dy, x, scale, mean, var
            )
        else:
            dx, ds, db = singa.CpuBatchNormBackward(
                self.handle, dy, x, scale, mean, var
            )
            
        #return dx, ds, db
        return dx, ds, db, self.running_mean, self.running_var
```
and wish to collect it with
```
        #all reduce running mean and var
        for p, g in autograd.backward(loss):
            if((p.requires_grad==False) and (p.stores_grad==False)):
                all_reduce(p)
```

However, this is the error in return
```
Traceback (most recent call last):
  File "resnet_multigpu.py", line 163, in <module>
    for p, g in autograd.backward(loss):
  File "/usr/local/lib/python3.5/dist-packages/singa/autograd.py", line 136, in backward
    % (len(op.src), len(dxs))
AssertionError: the number of src ops (=3) and dx (=5) not match
```